allennlp/modules/transformer/t5.py

```diff
@@ -773,187 +801,473 @@ def forward(
+        ]
+        return cls(token_embeddings, blocks, final_layer_norm=final_layer_norm, dropout=dropout)
 
 
 class T5DecoderStack(T5Stack, FromParams):
     def __init__(
         self,
-        layers: Optional[List[T5Block]] = None,
-        token_embeddings: Optional[nn.Embedding] = None,
+        token_embeddings: nn.Embedding,
+        blocks: List[T5Block],
         final_layer_norm: Optional[T5LayerNorm] = None,
         dropout: float = 0.1,
     ):
-        if layers is None:
-            layers = [
-                T5Block(
-                    attention=T5LayerSelfAttention(
-                        T5Attention(True, has_relative_attention_bias=(i == 0))
-                    ),
-                    cross_attention=T5LayerCrossAttention(),
-                )
-                for i in range(24)
-            ]
-        else:
-            if not all(b.is_decoder for b in layers):
-                from allennlp.common.checks import ConfigurationError
-
-                raise ConfigurationError(
-                    "Found an encoder block in a decoder stack. This won't work."
-                )
+        if not all(b.is_decoder for b in blocks):
+            raise ConfigurationError("Found an encoder block in a decoder stack. This won't work.")
+
+        super().__init__(
+            token_embeddings,
+            blocks,
+            final_layer_norm=final_layer_norm,
+            dropout=dropout,
+        )
 
-        super().__init__(layers, token_embeddings, final_layer_norm, dropout)
+    @classmethod
+    def basic_decoder(
+        cls,
+        token_embeddings: nn.Embedding,
+        num_blocks: int = 6,
+        block_self_attention: Lazy[T5Attention] = Lazy(T5Attention),
+        block_cross_attention: Lazy[T5Attention] = Lazy(T5Attention),
+        final_layer_norm: Optional[T5LayerNorm] = None,
+        block_ff: Lazy[T5LayerFF] = Lazy(T5LayerFF),
+        dropout: float = 0.1,
+    ) -> "T5DecoderStack":
+        blocks = [
+            T5Block(
+                attention=T5LayerSelfAttention(
+                    self_attention=block_self_attention.construct(
+                        is_decoder=True, has_relative_attention_bias=(i == 0)
+                    )
+                ),
+                cross_attention=T5LayerCrossAttention(
+                    enc_dec_attention=block_cross_attention.construct(
+                        is_decoder=True,
+                        has_relative_attention_bias=False,
+                    )
+                ),
+                ff=block_ff.construct(),
+            )
+            for i in range(num_blocks)
+        ]
+        return cls(token_embeddings, blocks, final_layer_norm=final_layer_norm, dropout=dropout)
 
 
 @dataclass
-class T5ForConditionalGenerationOutput:
-    logits: FloatT
-    loss: Optional[FloatT] = None
-    past_key_values: Optional[List[KeyValueStates]] = None
-    decoder_hidden_states: Optional[List[FloatT]] = None
+class T5Output:
```

Comment:
Even if we do nothing else, this version of T5, with the types and the docs and everything, is already a superior starting point if you want to hack on top of T5. I wonder if we should see if HF would want some if this ported back? Let's talk about that once this PR is done.